modules/afsql/afsql.c

```diff
@@ -805,7 +806,73 @@ afsql_dd_build_insert_command(AFSqlDestDriver *self, LogMessage *msg, GString *t
             }
           else
             {
-              dbi_conn_quote_string_copy(self->dbi_ctx, value->str, &quoted);
+              TypeHint type = self->fields[i].value->type_hint;
+              switch(type)
+                {
+                case TYPE_HINT_INT32:
+                case TYPE_HINT_INT64:
+                {
+                  gint64 k;
+                  if (type_cast_to_int64(value->str, &k, NULL))
+                    {
+                      dbi_conn_escape_string_copy(self->dbi_ctx, value->str, &quoted);
+                    }
+                  else
+                    {
+                      type_cast_drop_helper(self->template_options.on_error,
+                                            value->str, "int");
+                      if(fallback)
+                        {
+                          dbi_conn_quote_string_copy(self->dbi_ctx, value->str, &quoted);
+                        }
+                    }
+                  break;
```

Comment:
each of these blocks should have the following outcomes:

1) value is ok, should be added to the INSERT statement as a number (and not as a string)
2) value is not ok, fallback to string is TRUE, it should be added to the INSERT statement as a string.
3) value is not ok, drop is enabled, then the entire INSERT statement should be dropped instead of just the column

By looking at these case statements, I can see the following issues:
1) you encode the value as a string (by using the return value of dbi_conn_escape_string_copy) and not as a number literal.
2) you don't drop the entire INSERT statement if type_cast_drop_helper() returns TRUE (which you should)
3) the fallback seems ok.

Reply:
@bazsi, Regarding the first point, I couldn't find a function to encode the value as a number literal. Hence, I used an escape_string which is basically a string without double-quotes. http://libdbi.sourceforge.net/docs/programmers-guide/reference-query.html#DBI-CONN-QUERY
Am I missing something?